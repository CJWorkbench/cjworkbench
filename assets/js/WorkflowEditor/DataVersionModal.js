import React from 'react'
import PropTypes from 'prop-types'
import { Modal, ModalHeader, ModalBody, ModalFooter } from '../components/Modal'
import memoize from 'memoize-one'
import { setDataVersionAction, setWfModuleNotificationsAction } from '../workflow-reducer'
import { connect } from 'react-redux'
import { createSelector } from 'reselect'
import { Trans } from '@lingui/macro'

const Months = [
  // AP style
  'Jan.',
  'Feb.',
  'March',
  'April',
  'May',
  'June',
  'July',
  'Aug.',
  'Sept.',
  'Oct.',
  'Nov.',
  'Dec.'
]

// Always print as if our time zone is UTC, when testing
// (all other solutions are worse, including env vars and pre-adjusted test data)
let _formatDateUTCforTesting = false
export function formatDateUTCForTesting () {
  _formatDateUTCforTesting = true
}

/**
 * Return AP-style date in the user's time zone, but with the date before the
 * time.
 *
 * For instance: "June 22, 2018 – 10:35 a.m."
 */
function formatDate (date) {
  let mon, dd, yyyy, hh, mm
  if (_formatDateUTCforTesting) {
    mon = Months[date.getUTCMonth()]
    dd = date.getUTCDate()
    yyyy = date.getUTCFullYear()
    hh = date.getUTCHours()
    mm = String(100 + date.getUTCMinutes()).slice(1) // 0-padded
  } else {
    mon = Months[date.getMonth()]
    dd = date.getDate()
    yyyy = date.getFullYear()
    hh = date.getHours()
    mm = String(100 + date.getMinutes()).slice(1) // 0-padded
  }
  let ampm = 'a.m.'

  if (hh === 0) {
    hh = 12
  } else if (hh === 12) {
    ampm = 'p.m.'
  } else if (hh > 12) {
    ampm = 'p.m.'
    hh -= 12
  }

  return `${mon} ${dd}, ${yyyy} – ${hh}:${mm} ${ampm}`
}

/**
 * Form <input type="radio">. Calls onSelect on change.
 */
class FetchVersion extends React.PureComponent {
  static propTypes = {
    id: PropTypes.string.isRequired, // version ID
    date: PropTypes.instanceOf(Date).isRequired, // version date
    isSelected: PropTypes.bool.isRequired,
    isSeen: PropTypes.bool.isRequired, // has user selected this version ever
    onSelect: PropTypes.func.isRequired // func(versionId) => undefined
  }

  handleChange = (ev) => {
    if (ev.target.checked) {
      this.props.onSelect(this.props.id)
    }
  }

  render () {
    const { id, date, isSeen, isSelected } = this.props

    let className = isSeen ? 'seen' : 'unseen'
    if (isSelected) className += ' selected'

    return (
      <label className={className}>
        <input type='radio' name='data-version' value={id} checked={isSelected} onChange={this.handleChange} />
        <time time={date.toISOString()}>{formatDate(date)}</time>
      </label>
    )
  }
}

/**
 * Form that calls onSubmit(wantNotifications).
 *
 * Implemented as a checkbox with an onChange event. ([2018-06-22] it _looks_
 * like a <button>, but it's a checkbox in a <label>.)
 */
class NotificationsForm extends React.PureComponent {
  static propTypes = {
    notificationsEnabled: PropTypes.bool.isRequired,
    onSubmit: PropTypes.func.isRequired // func(bool) => undefined
  }

  handleChange = (ev) => {
    this.props.onSubmit(ev.target.checked) // should change the state
  }

  handleSubmit = (ev) => {
    ev.preventDefault()
    ev.stopPropagation()
  }

  render () {
    const checked = this.props.notificationsEnabled
    const className = checked ? 'notifications-enabled' : 'notifications-disabled'
    const iconAlert = checked ? 'icon-notification' : 'icon-no-notification'

    return (
      <form onSubmit={this.handleSubmit} className={`notifications ${className}`}>
        <div className='text'>
          <p className='status'><i className={`icon ${iconAlert}`} />{
            checked
              ? <Trans id='js.WorkflowEditor.DataVersionModal.NotificationsForm.status.alertsOn' description='The tag adds emphasis'>Alerts are <strong>on</strong></Trans>
              : <Trans id='js.WorkflowEditor.DataVersionModal.NotificationsForm.status.alertsOff' description='The tag adds emphasis'>Alerts are <strong>off</strong></Trans>
          }
          </p>
          <p className='description'>{
            checked
              ? <Trans id='js.WorkflowEditor.DataVersionModal.NotificationsForm.description.emailOnOutputChange'>You will receive and email if the output of this module changes</Trans>
              : <Trans id='js.WorkflowEditor.DataVersionModal.NotificationsForm.description.turnAlertsOn'>Turn alerts ON to receive an email if the output of this module changes</Trans>
          }
          </p>
        </div>
        <div className='options'>
          <label>
            <input name='notifications-enabled' type='checkbox' checked={checked} onChange={this.handleChange} />
            <span className='action'>{checked ? <Trans id='js.WorkflowEditor.DataVersionModal.NotificationsForm.options.turnOff'>Turn off</Trans> : <Trans id='js.WorkflowEditor.DataVersionModal.NotificationsForm.options.turnOn'>Turn on</Trans>}</span>
          </label>
        </div>
      </form>
    )
  }
}

export class DataVersionModal extends React.PureComponent {
  static propTypes = {
    fetchWfModuleId: PropTypes.number.isRequired,
    fetchVersions: PropTypes.arrayOf(PropTypes.shape({
      id: PropTypes.string.isRequired,
      date: PropTypes.instanceOf(Date).isRequired,
      isSeen: PropTypes.bool.isRequired
    })).isRequired,
    selectedFetchVersionId: PropTypes.string, // null for no selection
    wfModuleId: PropTypes.number.isRequired,
    isAnonymous: PropTypes.bool.isRequired,
    notificationsEnabled: PropTypes.bool.isRequired, // whether enabled on selectedWfModule
    onClose: PropTypes.func.isRequired, // func() => undefined
    onChangeFetchVersionId: PropTypes.func.isRequired, // func(wfModuleId, versionId) => undefined
    onChangeNotificationsEnabled: PropTypes.func.isRequired // func(wfModuleId, isEnabled) => undefined
  }

  state = {
    selectedFetchVersionId: this.props.selectedFetchVersionId
  }

  handleChangeNotificationsEnabled = (isEnabled) => {
    const { onChangeNotificationsEnabled, wfModuleId } = this.props
    onChangeNotificationsEnabled(wfModuleId, isEnabled)
  }

  handleSelectSelectedFetchVersionId = (selectedFetchVersionId) => {
    this.setState({ selectedFetchVersionId })
  }

  handleSubmit = () => {
    if (this.state.selectedFetchVersionId !== this.props.selectedFetchVersionId) {
      const { fetchWfModuleId, onChangeFetchVersionId } = this.props
      onChangeFetchVersionId(fetchWfModuleId, this.state.selectedFetchVersionId)
    }

    this.props.onClose()
  }

  render () {
    const {
      fetchVersions,
      onClose,
      isAnonymous,
      notificationsEnabled
    } = this.props

    return (
      <Modal className='data-versions-modal' isOpen fade={false} toggle={onClose}>
        <ModalHeader toggle={onClose}>Data Versions</ModalHeader>
        <ModalBody>
          <form onSubmit={this.handleSubmit} onCancel={onClose}>
            <ol>
              {fetchVersions.map(v => (
                <li key={v.id}>
                  <FetchVersion
                    onSelect={this.handleSelectSelectedFetchVersionId}
                    isSelected={this.state.selectedFetchVersionId === v.id}
                    {...v}
                  />
                </li>
              ))}
            </ol>
          </form>
        </ModalBody>
        <ModalFooter>
          {isAnonymous ? null : (
            <NotificationsForm
              notificationsEnabled={notificationsEnabled}
              onSubmit={this.handleChangeNotificationsEnabled}
            />
          )}
          <div className='actions'>
            <button
              name='load'
              disabled={this.state.selectedFetchVersionId === this.props.selectedFetchVersionId}
              onClick={this.handleSubmit}
            >Load
            </button>
          </div>
        </ModalFooter>
      </Modal>
    )
  }
}

const getWorkflow = ({ workflow }) => workflow
const getTabs = ({ tabs }) => tabs
const getSelectedTab = createSelector([getWorkflow, getTabs], (workflow, tabs) => {
  return tabs[workflow.tab_slugs[workflow.selected_tab_position]]
})
const getWfModules = ({ wfModules }) => wfModules
const getSelectedTabWfModules = createSelector([getSelectedTab, getWfModules], (tab, wfModules) => {
  return tab.wf_module_ids.map(id => wfModules[String(id)] || null)
})
const getModules = ({ modules }) => modules
/**
 * Find first (WfModule, Module) that has a `.loads_data` ModuleVersion.
 */
const getFetchWfModule = createSelector([getSelectedTabWfModules, getModules], (wfModules, modules) => {
  for (const wfModule of wfModules) {
    const module = modules[wfModule.module] || {}
    if (module.loads_data) {
      return wfModule
    }
  }

  return null
})

/**
 * Parse `wfModule.versions.versions` Array of { id, date, isSeen }.
 *
 * wfModule.versions.versions is an Array of [ dateString, isSeen ]
 * pairs.
 */
const getFetchVersions = memoize(versions => {
  return versions.map(version => {
    const [id, isSeen] = version
    return {
      id,
      isSeen,
      date: new Date(id)
    }
  })
})

function mapStateToProps (state, { wfModuleId }) {
  const fetchWfModule = getFetchWfModule(state)
  const wfModule = state.wfModules[String(wfModuleId)]
  const notificationsEnabled = wfModule ? wfModule.notifications : false

  return {
    fetchWfModuleId: fetchWfModule ? fetchWfModule.id : null,
    fetchVersions: fetchWfModule ? getFetchVersions(fetchWfModule.versions.versions || []) : null,
    selectedFetchVersionId: fetchWfModule ? fetchWfModule.versions.selected : null,
    isAnonymous: state.workflow.is_anonymous,
    notificationsEnabled
  }
}

const mapDispatchToProps = {
  onChangeFetchVersionId: setDataVersionAction,
  onChangeNotificationsEnabled: setWfModuleNotificationsAction
}

export default connect(
  mapStateToProps,
  mapDispatchToProps
)(DataVersionModal)
