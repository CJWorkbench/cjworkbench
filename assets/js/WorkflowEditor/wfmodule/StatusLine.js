// ---- StatusLine ----

// Display error message, if any
// BUG - Tying this to Props will ensure that error message stays displayed, even after resolution
import React, { useCallback, useEffect, useState } from 'react'
import PropTypes from 'prop-types'
import QuickFix, { QuickFixPropTypes } from './QuickFix'

const StatusLine = React.memo(function StatusLine ({ status, errors, applyQuickFix, isReadOnly }) {
  const [clickedAnyQuickFix, setClickedQuickFix] = useState(false)
  const doApplyQuickFix = useCallback((...args) => {
    setClickedQuickFix(true)
    applyQuickFix(...args)
  })

  // after props change (remember: we're in React.memo), assume the quick fix
  // suggestions are not-yet-clicked.
  useEffect(() => setClickedQuickFix(false), [status, errors])

  if (!errors.length) return null

  return (
    <>
      {errors.map(({ message, quickFixes }, j) => (
        <div className='wf-module-error-msg' key={j}>
          <p>{message}</p>
          {quickFixes && quickFixes.length && !isReadOnly ? (
            <ul className='quick-fixes'>
              {quickFixes.map((quickFix, i) => (
                <li key={i}>
                  <QuickFix
                    disabled={clickedAnyQuickFix}
                    applyQuickFix={doApplyQuickFix}
                    {...quickFix}
                  />
                </li>
              ))}
            </ul>
          ) : null}
        </div>
      ))}
    </>
  )
})
StatusLine.propTypes = {
  status: PropTypes.oneOf(['ok', 'busy', 'error', 'unreachable']).isRequired,
  isReadOnly: PropTypes.bool.isRequired, // if true, cannot apply quick fixes
  errors: PropTypes.arrayOf(PropTypes.shape({ message: PropTypes.string.isRequired, quickFixes: PropTypes.arrayOf(PropTypes.shape(QuickFixPropTypes)) })), // may be empty
  applyQuickFix: PropTypes.func.isRequired // func(action, args) => undefined
}

export default StatusLine
