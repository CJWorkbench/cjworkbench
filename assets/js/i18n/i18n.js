import { setupI18n } from "@lingui/core"
import { currentLocale } from './locales'
import { fetchCatalog } from './catalogs'

export default const i18n = setupI18n({
   language: currentLocale,
   catalogs: {
       currentLocale: fetchCatalog(currentLocale)
   }
})
