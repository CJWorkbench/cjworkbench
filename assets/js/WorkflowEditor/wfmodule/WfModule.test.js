/* globals describe, it, expect, jest, beforeEach, afterEach */
import React from 'react'
import { act } from 'react-dom/test-utils'
import ConnectedWfModule, { WfModule } from './WfModule'
import DataVersionModal from '../DataVersionModal'
import { mockStore } from '../../test-utils'
import { shallowWithI18n, mountWithI18n } from '../../i18n/test-utils'
import deepEqual from 'fast-deep-equal'
import { createStore } from 'redux'
import { Provider } from 'react-redux'

import { generateSlug } from '../../utils'
import lessonSelector from '../../lessons/lessonSelector'
import StatusLine from './StatusLine'

jest.mock('../../utils')
jest.mock('../../lessons/lessonSelector', () => jest.fn()) // same mock in every test :( ... we'll live

describe('WfModule, not read-only mode', () => {
  let mockApi

  beforeEach(() => {
    mockApi = {
      createOauthAccessToken: jest.fn(),
      valueCounts: jest.fn()
    }
  })

  // A mock module that looks like LoadURL
  const wfModule = {
    id: 999,
    slug: 'step-1',
    notes: '',
    is_collapsed: false, // false because we render more, so better test
    is_busy: false,
    output_status: 'ok',
    params: {
      url: 'http://some.URL.me',
      menu_select: 1
    },
    secrets: {},
    files: [],
    output_errors: []
  }

  function pspec (idName, type, extraProps = {}) {
    return {
      idName,
      name: '',
      type,
      multiline: false,
      placeholder: '',
      visibleIf: null,
      ...extraProps
    }
  }

  const module = {
    id_name: 'loadurl', // more or less
    name: 'Load from URL',
    icon: 'icon',
    help_url: 'http://help-url',
    param_fields: [
      pspec('url', 'string'),
      pspec('menu_select', 'menu', { items: 'Mango|Banana' })
    ]
  }

  const wrapper = (extraProps = {}) => {
    return shallowWithI18n(
      <WfModule
        isReadOnly={false}
        isAnonymous={false}
        isZenMode={false}
        isZenModeAllowed={false}
        module={module}
        workflowId={99}
        wfModule={wfModule}
        removeModule={jest.fn()}
        inputWfModule={{ id: 123, last_relevant_delta_id: 707 }}
        isSelected
        isAfterSelected={false}
        api={mockApi}
        tabs={[]}
        currentTab='tab-1'
        index={2}
        onDragStart={jest.fn()}
        onDragEnd={jest.fn()}
        isLessonHighlight={false}
        isLessonHighlightCollapse={false}
        isLessonHighlightNotes={false}
        fetchModuleExists={false}
        clearNotifications={jest.fn()}
        setSelectedWfModule={jest.fn()}
        setWfModuleCollapsed={jest.fn()}
        setWfModuleParams={jest.fn()}
        setWfModuleNotes={jest.fn()}
        maybeRequestFetch={jest.fn()}
        setZenMode={jest.fn()}
        applyQuickFix={jest.fn()}
        {...extraProps}
      />
    )
  }

  it('matches snapshot', () => {
    expect(wrapper()).toMatchSnapshot()
  })

  it('is has .status-busy', () => {
    const w = wrapper({ wfModule: { ...wfModule, output_status: 'busy' } })
    expect(w.hasClass('status-busy')).toBe(true)
    expect(w.find('ParamsForm').prop('isWfModuleBusy')).toBe(true)
    expect(w.find(StatusLine).prop('status')).toEqual('busy')

    w.setProps({ wfModule: { ...wfModule, output_status: 'ok' } })
    w.update()
    expect(w.hasClass('status-busy')).toBe(false)
    expect(w.hasClass('status-ok')).toBe(true)

    expect(w.find('ParamsForm').prop('isWfModuleBusy')).toBe(false)
    expect(w.find(StatusLine).prop('status')).toEqual('ok')
    expect(w.hasClass('status-ok')).toBe(true)
    expect(w.hasClass('status-busy')).toBe(false)
  })

  it('does not render ParamsForm when collapsed', () => {
    // https://www.pivotaltracker.com/story/show/165539156
    // If we were to render a <ParamsForm> while display:none was set, the
    // params wouldn't be able to auto-size themselves.
    const w = wrapper({ wfModule: { ...wfModule, is_collapsed: true } })
    expect(w.find('ParamsForm').length).toEqual(0)
  })

  it('has .status-busy overridden when wfModule.is_busy', () => {
    const w = wrapper({ wfModule: { ...wfModule, output_status: 'ok', is_busy: true } })
    expect(w.hasClass('status-busy')).toBe(true)
  })

  it('renders a note', () => {
    const w = wrapper({ wfModule: { ...wfModule, notes: 'some notes' } })
    expect(w.find('EditableNotes').prop('value')).toEqual('some notes')
  })

  it('renders in Zen mode', () => {
    const w = wrapper({ isZenMode: true })
    expect(w.prop('className')).toMatch(/\bzen-mode\b/)
    expect(w.find('ParamsForm').prop('isZenMode')).toBe(true)
  })

  it('has an "enter zen mode" button', () => {
    const w = wrapper({ isZenModeAllowed: true, isZenMode: false })
    let checkbox = w.find('input[type="checkbox"][name="zen-mode"]')
    expect(checkbox.prop('checked')).toBe(false)
    checkbox.simulate('change', { target: { checked: true } })
    expect(w.instance().props.setZenMode).toHaveBeenCalledWith(wfModule.id, true)
    w.setProps({ isZenMode: true })

    checkbox = w.find('input[type="checkbox"][name="zen-mode"]')
    expect(checkbox.prop('checked')).toBe(true)
    checkbox.simulate('change', { target: { checked: false } })
    expect(w.instance().props.setZenMode).toHaveBeenCalledWith(wfModule.id, false)
  })

  it('renders notifications, opening and closing a modal', () => {
    const w = wrapper({ fetchModuleExists: true, isReadOnly: false, isAnonymous: false })

    w.find('button.notifications').simulate('click')
    expect(w.instance().props.clearNotifications).toHaveBeenCalledWith(999)
    expect(w.find(DataVersionModal).length).toBe(1)

    w.find(DataVersionModal).prop('onClose')()
    w.update()
    expect(w.find(DataVersionModal).length).toBe(0)
  })

  it('hides notifications when isAnonymous', () => {
    const w = wrapper({ fetchModuleExists: true, isReadOnly: false, isAnonymous: true })
    expect(w.find('button.notifications').length).toBe(0)
  })

  it('hides notifications when isReadOnly', () => {
    const w = wrapper({ fetchModuleExists: true, isReadOnly: true, isAnonymous: false })
    expect(w.find('button.notifications').length).toBe(0)
  })

  it('hides notifications when there is no fetch module', () => {
    const w = wrapper({ fetchModuleExists: false, isReadOnly: false, isAnonymous: false })
    expect(w.find('button.notifications').length).toBe(0)
  })

  it('adds a note', () => {
    const w = wrapper({ wfModule: { ...wfModule, notes: '' } })

    expect(w.find('.module-notes.visible')).toHaveLength(0)

    w.find('button.edit-note').simulate('click')
    w.find('EditableNotes').simulate('change', { target: { value: 'new note' } })
    w.find('EditableNotes').simulate('blur')

    expect(w.instance().props.setWfModuleNotes).toHaveBeenCalledWith(wfModule.id, 'new note')
  })

  it('deletes a note', () => {
    const w = wrapper({ wfModule: { ...wfModule, notes: 'some notes' } })

    expect(w.find('.module-notes.visible')).toHaveLength(1)

    w.find('EditableNotes').simulate('change', { target: { value: '' } })
    w.find('EditableNotes').simulate('blur')

    expect(w.instance().props.setWfModuleNotes).toHaveBeenCalledWith(wfModule.id, '')
  })

  it('does not show old value during edit when editing ""', () => {
    // https://www.pivotaltracker.com/story/show/163005781
    const w = wrapper({ wfModule: { ...wfModule, notes: 'some notes' } })

    w.find('EditableNotes').simulate('change', { target: { value: '' } })
    expect(w.find('EditableNotes').prop('value')).toEqual('')
  })

  it('queues changes from onChange and then submits them in onSubmit', () => {
    const wfModule = {
      id: 999,
      notes: '',
      is_collapsed: false,
      params: {
        a: 'A',
        b: 'B'
      },
      secrets: {},
      files: [],
      output_errors: []
    }
    const aModule = {
      ...module,
      param_fields: [pspec('a', 'string'), pspec('b', 'string')]
    }
    // Make setWfModuleParams() return a thennable -- a fake promise.
    // It'll set onSetWfModuleParamsDone, which we can call synchronously.
    // (This is white-box testing -- we assume the .then() and the sync
    // call.)
    let onSetWfModuleParamsDone = null
    const setWfModuleParams = jest.fn(() => ({ then: (fn) => { onSetWfModuleParamsDone = fn } }))
    const w = wrapper({
      wfModule,
      module: aModule,
      setWfModuleParams
    })

    w.find('ParamsForm').prop('onChange')({ a: 'C' })
    w.update()
    expect(w.find('ParamsForm').prop('edits')).toEqual({ a: 'C' })

    // not submitted to the server
    expect(setWfModuleParams).not.toHaveBeenCalled()
    expect(w.prop('className')).toMatch(/\bediting\b/)

    w.find('ParamsForm').prop('onSubmit')()
    expect(setWfModuleParams).toHaveBeenCalledWith(999, { a: 'C' })

    // a bit of a white-box test: state should be cleared ... but only
    // after a tick, so Redux can do its magic in setWfModuleParams
    expect(w.prop('className')).toMatch(/\bediting\b/)
    expect(w.state('edits')).toEqual({ a: 'C' })
    w.setProps({ wfModule: { ...wfModule, params: { a: 'C' } } })
    act(onSetWfModuleParamsDone)
    expect(w.state('edits')).toEqual({})
    expect(w.prop('className')).not.toMatch(/\bediting\b/)
  })

  it('submits a fetch event in onSubmit', () => {
    // Use case:
    // 1. User edits url field
    // 2. User clicks "submit" button
    const wfModule = {
      id: 999,
      notes: '',
      is_collapsed: false,
      params: {
        url: '',
        version_select: 'B'
      },
      secrets: {},
      files: [],
      output_errors: []
    }
    const setWfModuleParams = jest.fn(() => ({ then: jest.fn() })) // dummy promise
    const aModule = {
      ...module,
      param_fields: [pspec('url', 'string'), pspec('version_select', 'custom')]
    }
    const w = wrapper({ wfModule, module: aModule, setWfModuleParams })

    w.find('ParamsForm').prop('onChange')({ url: 'http://example.org' })
    w.find('ParamsForm').prop('onSubmit')()

    expect(setWfModuleParams).toHaveBeenCalledWith(999, { url: 'http://example.org' })
    expect(w.instance().props.maybeRequestFetch).toHaveBeenCalledWith(999)
  })

  it('overrides status to busy when a fetch is pending', () => {
    const w = wrapper({ wfModule: { ...wfModule, nClientRequests: 1 } })
    expect(w.find('ParamsForm').prop('isWfModuleBusy')).toBe(true)
    expect(w.find(StatusLine).prop('status')).toEqual('busy')
    expect(w.prop('className')).toMatch(/\bstatus-busy\b/)
  })

  it('applies a quick fix', () => {
    // Scenario: user is on linechart and chose non-numeric Y axis
    mockApi.setSelectedWfModule = jest.fn(() => Promise.resolve(null))
    mockApi.addModule = jest.fn(() => Promise.resolve(null))
    generateSlug.mockImplementation(prefix => prefix + 'X')
    const quickFix = {
      buttonText: 'Fix it',
      action: { type: 'prependStep', moduleSlug: 'dosomething', partialParams: { A: 'B' } }
    }
    const store = mockStore({
      workflow: {
        id: 99,
        tab_slugs: ['tab-11', 'tab-12'],
        read_only: false,
        is_anonymous: false,
        selected_tab_position: 0
      },
      tabs: {
        'tab-11': { slug: 'tab-11', name: 'Tab 1', wf_module_ids: [10, 20] },
        'tab-12': { slug: 'tab-12', name: 'Tab 2', wf_module_ids: [] }
      },
      wfModules: {
        10: { id: 10, slug: 'step-10', tab_slug: 'tab-11' },
        20: { id: 20, slug: 'step-20', tab_slug: 'tab-11' }
      },
      modules: {
        loadurl: {
          name: 'Load from URL',
          id_name: 'loadurl',
          help_url: '',
          icon: 'icon',
          param_fields: []
        }
      }
    }, mockApi)

    lessonSelector.mockReset()
    lessonSelector.mockReturnValue({
      testHighlight: _ => false
    })

    const w = mountWithI18n(
      <Provider store={store}>
        <ConnectedWfModule
          isReadOnly={false}
          isAnonymous={false}
          isZenMode={false}
          isZenModeAllowed={false}
          index={1}
          wfModule={{ id: 20, module: 'loadurl', is_collapsed: false, status: 'error', params: {}, secrets: {}, output_errors: [{ message: 'foo', quickFixes: [quickFix] }], files: [] }}
          isSelected
          isAfterSelected={false}
          onDragStart={jest.fn()}
          onDragEnd={jest.fn()}
          setZenMode={jest.fn()}
          api={mockApi}
        />
      </Provider>
    )

    w.find('button.quick-fix').simulate('click')
    expect(mockApi.addModule).toHaveBeenCalledWith('tab-11', 'step-X', 'dosomething', 1, { A: 'B' })
  })

  describe('lesson highlights', () => {
    let store
    let wrapper = null
    let nonce = 0

    function highlight (selectors) {
      lessonSelector.mockReturnValue({
        testHighlight: test => {
          return selectors.some(s => deepEqual(test, s))
        }
      })

      // trigger a change
      store.dispatch({ type: 'whatever', payload: ++nonce })
      if (wrapper !== null) wrapper.update()
    }

    beforeEach(() => {
      lessonSelector.mockReset()

      store = createStore((_, action) => ({
        workflow: {
          id: 99,
          read_only: false,
          is_anonymous: false,
          tab_slugs: ['tab-11'],
          selected_tab_position: 0
        },
        tabs: {
          'tab-11': { slug: 'tab-11', name: 'Tab 1', wf_module_ids: [1, 2, 999] }
        },
        wfModules: {
          999: { slug: 'step-99', module: 'loadurl', params: {}, secrets: {} }
        },
        modules: {
          loadurl: {
            id_name: 'loadurl',
            name: 'Load from URL',
            help_url: '',
            icon: 'icon',
            param_fields: [pspec('url', 'string')]
          }
        },
        ...action.payload
      }))

      highlight([])

      wrapper = mountWithI18n(
        <Provider store={store}>
          <ConnectedWfModule
            isReadOnly={false}
            isAnonymous={false}
            isZenMode={false}
            isZenModeAllowed={false}
            index={1}
            wfModule={{ id: 20, module: 'loadurl', is_collapsed: false, status: 'error', params: {}, secrets: {}, output_errors: [{ message: 'foo', quickFixes: [] }], files: [] }}
            isSelected
            isAfterSelected={false}
            onDragStart={jest.fn()}
            onDragEnd={jest.fn()}
            setZenMode={jest.fn()}
            api={mockApi}
          />
        </Provider>
      )
    })
    afterEach(() => {
      if (wrapper !== null) {
        wrapper.unmount()
        wrapper = null
      }
    })

    it('highlights a WfModule', () => {
      highlight([{ type: 'WfModule', index: 1, moduleName: 'Load from URL' }])
      expect(wrapper.find('.wf-module').prop('className')).toMatch(/\blesson-highlight\b/)
    })

    it('unhighlights a WfModule', () => {
      // wrong name
      highlight([{ type: 'WfModule', index: 1, moduleName: 'TestModule2' }])
      expect(wrapper.find('.wf-module').prop('className')).not.toMatch(/\blesson-highlight\b/)
    })

    it('highlights the "collapse" button', () => {
      highlight([{ type: 'WfModuleContextButton', index: 1, moduleName: 'Load from URL', button: 'collapse' }])
      expect(wrapper.find('i.context-collapse-button').prop('className')).toMatch(/\blesson-highlight\b/)
    })

    it('unhighlights the "collapse" button', () => {
      // wrong moduleName
      highlight([{ type: 'WfModuleContextButton', index: 1, moduleName: 'TestModule2', button: 'collapse' }])
      expect(wrapper.find('i.context-collapse-button').prop('className')).not.toMatch(/\blesson-highlight\b/)

      // wrong button
      highlight([{ type: 'WfModuleContextButton', index: 1, moduleName: 'Load from URL', button: 'notes' }])
      expect(wrapper.find('i.context-collapse-button').prop('className')).not.toMatch(/\blesson-highlight\b/)
    })

    it('highlights the notes button', () => {
      highlight([{ type: 'WfModuleContextButton', index: 1, moduleName: 'Load from URL', button: 'notes' }])
      expect(wrapper.find('button.edit-note').prop('className')).toMatch(/\blesson-highlight\b/)
    })

    it('unhighlights the notes button', () => {
      // wrong moduleName
      highlight([{ type: 'WfModuleContextButton', index: 1, moduleName: 'TestModule2', button: 'notes' }])
      expect(wrapper.find('button.edit-note').prop('className')).not.toMatch(/\blesson-highlight\b/)
    })
  })
})
