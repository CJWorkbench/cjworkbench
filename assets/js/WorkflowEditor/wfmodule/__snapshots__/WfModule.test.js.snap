// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`WfModule, not read-only mode matches snapshot 1`] = `
<div
  className="wf-module status-ok selected"
  data-module-name="Load from URL"
  data-module-slug="loadurl"
  onMouseDown={[Function]}
>
  <div
    className="module-notes"
  >
    <EditableNotes
      inputRef={
        Object {
          "current": null,
        }
      }
      isReadOnly={false}
      onBlur={[Function]}
      onCancel={[Function]}
      onChange={[Function]}
      onFocus={[Function]}
      placeholder="Type a note..."
      value=""
    />
  </div>
  <h3>
    3
  </h3>
  <div
    className="module-card-and-link"
  >
    <div
      className="module-card"
      draggable={true}
      onDragEnd={[Function]}
      onDragStart={[Function]}
    >
      <div
        className="module-card-header"
      >
        <div
          className="controls"
        >
          <WfModuleCollapseButton
            isCollapsed={false}
            isLessonHighlight={false}
            onCollapse={[Function]}
            onExpand={[Function]}
          />
          <i
            className="icon-icon WFmodule-icon"
          />
          <div
            className="module-name"
          >
            Load from URL
          </div>
          <div
            className="context-buttons"
          >
            <a
              className="help-button"
              href="http://help-url"
              rel="noopener noreferrer"
              target="_blank"
              title="Help for this module"
            >
              <i
                className="icon-help"
              />
            </a>
            <button
              className="btn edit-note"
              onClick={[Function]}
              title="Edit Note"
            >
              <i
                className="icon-note"
              />
            </button>
            <WithI18n
              id={999}
              removeModule={[Function]}
            />
          </div>
        </div>
        <DeprecationMessage
          helpUrl="http://help-url"
          message={null}
        />
      </div>
      <div
        className="module-card-details expanded"
      >
        <Memo(StatusLine)
          applyQuickFix={[Function]}
          errors={Array []}
          isReadOnly={false}
          module={
            Object {
              "help_url": "http://help-url",
              "icon": "icon",
              "id_name": "loadurl",
              "name": "Load from URL",
              "param_fields": Array [
                Object {
                  "idName": "url",
                  "multiline": false,
                  "name": "",
                  "placeholder": "",
                  "type": "string",
                  "visibleIf": null,
                },
                Object {
                  "idName": "menu_select",
                  "items": "Mango|Banana",
                  "multiline": false,
                  "name": "",
                  "placeholder": "",
                  "type": "menu",
                  "visibleIf": null,
                },
              ],
            }
          }
          status="ok"
        />
        <ErrorBoundary>
          <ParamsForm
            api={
              Object {
                "createOauthAccessToken": [MockFunction],
                "valueCounts": [MockFunction],
              }
            }
            applyQuickFix={[Function]}
            currentTab="tab-1"
            deleteSecret={[Function]}
            edits={Object {}}
            fields={
              Array [
                Object {
                  "idName": "url",
                  "multiline": false,
                  "name": "",
                  "placeholder": "",
                  "type": "string",
                  "visibleIf": null,
                },
                Object {
                  "idName": "menu_select",
                  "items": "Mango|Banana",
                  "multiline": false,
                  "name": "",
                  "placeholder": "",
                  "type": "menu",
                  "visibleIf": null,
                },
              ]
            }
            files={Array []}
            inputDeltaId={null}
            inputWfModuleId={123}
            isReadOnly={false}
            isWfModuleBusy={false}
            isZenMode={false}
            onChange={[Function]}
            onSubmit={[Function]}
            secrets={Object {}}
            startCreateSecret={[Function]}
            submitSecret={[Function]}
            tabs={Array []}
            value={
              Object {
                "menu_select": 1,
                "url": "http://some.URL.me",
              }
            }
            wfModuleId={999}
            wfModuleOutputErrors={Array []}
            wfModuleSlug="step-1"
            workflowId={99}
          />
        </ErrorBoundary>
      </div>
    </div>
  </div>
</div>
`;
